feat(tags): adds `legacy_tag_formats` and `ignored_tag_formats` settings

noirbizarre opened the pull request:
<!--
Thanks for sending a pull request!
Please fill in the following content to let us know better about this change.
-->

## Description
<!-- Describe what the change is -->
This PR add 2 new settings to handle tag formats;
- `legacy_tag_formats` allowing tag format to change over time while retaining the ability to work with legacy version
- `ignored_tag_formats` to specify expected tag that should not raise `InvalidVersion`.

So projects which change tag formats and have expected but non-version tags can have these settings:

```toml
tag_format = "v${version}"
legacy_tag_formats = [
	"${version}",
	"project-prefix-${version}",
]
ignored_tag_formats = [
	"stable",
	"other-project-*",
]
```

By doing so, I had to start grouping tag handling and introduce a `tags.TagRules` object which is responsible for handling all tags rules, initially spread over `tag_format`, `version_scheme`s and `version_provider`s

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior
<!-- A clear and concise description of what you expected to happen -->

- I can change tag format and still being able to:
	- bump versions (aka. find the increment and commits range)
	- generate changelog (full, incremental, version range)
- I can generate a changelog without warnings for expected tags

## Steps to Test This Pull Request
- create a project
- set `version_provider = "scm"` (not mandatory but allows testing the full scope)
- bump some versions with an initial `tag_format`
- move current `tag_format` in `legacy_tag_format` and define a new `tag_format`
- bump a version with changelog:
	- current version should be properly identified
	- increment should properly been found
	- new version should be correct
	- generated changelog generation should be correct (both new and old version formats)
	- new tag should respect the new `tag_format`

then:
- add `random-*` to `ignored_tag_formats`
- tag any commit with `random-tag`
- run cz changelog
	- there shouldn't be any warning

## Additional context
<!-- Add any other RELATED ISSUE, context or screenshots about the pull request here. -->
There might be a 2nd PR handling a `changelog_version_format` setting to customize the version rendering in the changelog so:
- it's not the tag format anymore (problematic with some prefixes, like for monorepos)
- have a uniform version display across `tag_format` changes

Lee-W commented:
wow, this is a huge one. I'll be mostly out for the following month but will try to take a look after I'm back

noirbizarre commented:
Yes, I know, sorry. I had this idea long ago, just not having time.
Actually, this is as of today the main remaining pain point I have. I had to migrate a lot of projects on different tagging conventions in a corporate environment and as of today it involves lots of manual steps as well as features lost. 
So 🎉 there it is, this one fixes all my cases (and I'm sure I'm not the only one facing those).

In the meantime, enjoy your holidays !

woile commented:
Could you add a tutorial for this case? Otherwise it feels like we are adding a lot of features and it's unclear what a user can do with them.
I'll review it this week 👍🏻

noirbizarre commented:
I completed the monorepo tutorial and added a FAQ entry but yes, I can totally write a full tutorial on changing tag convention while keeping backward compatibility 👍🏽

Lee-W commented on commitizen/tags.py:
```diff
@@ -0,0 +1,201 @@
+        regexes = (self._format_regex(p, star=True) for p in self.ignored_tag_formats)
+        return [re.compile(r) for r in regexes]
+
+    def _format_regex(self, tag_pattern: str, star: bool = False) -> str:
```
What does `star` mean? Could we please add a docstring here? Thanks!

noirbizarre replied:
I'll add the missing docstrings 👍🏼

Lee-W commented on commitizen/tags.py:
```diff
@@ -0,0 +1,201 @@
+
+        return True
+
+    def search_version(self, txt: str, last: bool = False) -> tuple[str, str] | None:
```
the function name is a bit misleading. we probably should create 2 methods `search_first_version` and `serach_last_version` while using the same logic

noirbizarre replied:
They only differ by one line on 21, but I can add those as syntactic/semantic sugar over this one.

Lee-W reviewed (commented):
Hey @noirbizarre , thanks for making such an awesome PR! I left a few questions. I think we might need a few rounds of change and discussions before merging. But again, this is great!

noirbizarre commented on commitizen/version_schemes.py:
```diff
@@ -408,22 +407,22 @@ def __str__(self) -> str:
     :raises VersionSchemeUnknown: if the version scheme is not found.
     """
-    deprecated_setting: str | None = config.settings.get("version_type")
+    deprecated_setting: str | None = settings.get("version_type")
```
Agreed!

Lee-W replied:
maybe let's add a TODO comment here. WDYT?

noirbizarre replied:
Added 👍🏼 1

noirbizarre commented on commitizen/tags.py:
```diff
@@ -0,0 +1,201 @@
+            for f in (self.tag_format, *self.legacy_tag_formats)
+        ]
+        candidates = (t for t in tags if any(t.name == p for p in possible_tags))
+        return next(candidates, None)
```
Yes, because `possible_tags` are sorted (the current `tag_format` matches first) and we want the matching tag for a given version, there should be only one.
I'll add the missing docstrings

Lee-W replied:
if `there should be only one`, should we check if there're more and raise an error?

noirbizarre replied:
Added and tested 👍🏼

noirbizarre commented on commitizen/tags.py:
```diff
@@ -0,0 +1,201 @@
+
+        match = matches[-1 if last else 0]
+
+        if "version" in match.groupdict():
```
I tried, but they follow 2 different version/major/minor/patch extraction logic (I kept the existing logic)
But I'll see what I can do to handle this and the previous comment too

noirbizarre commented on commitizen/providers/scm_provider.py:
```diff
@@ -23,57 +14,14 @@ class ScmProvider(VersionProvider):
     It is meant for `setuptools-scm` or any package manager `*-scm` provider.
     """
 
-    TAG_FORMAT_REGEXS = get_tag_regexes(r"(?P<version>.+)")
```
It has moved into `tags.TagRules` too. It ensures that all tag rules are the same everywhere.
When you change the tag convention, the previous version needs to be matched against the legacy tag until a new version is tagged to the new format.

Lee-W replied:
This is niiice. Just want to be sure I did not miss anything

woile commented on commitizen/tags.py:
```diff
@@ -0,0 +1,214 @@
+
+
+@dataclass
+class TagRules:
```
could you add a mini-example of how to use this class? 🙏🏻 

```
@dataclass
class TagRules:
    """
    Encapsulate tag-related rules.

    Can some information come from the settings? is it design to give flexbility?
   
    Example:

    ```py
    rules = TagRules.from_settings(...)
    # etc...
    ```

Thanks 🙏🏻

noirbizarre replied:
I added a big one 👍🏼

woile commented:
I like this a lot 💪🏻 super cool work! if possible please add some tests for `test_tag` mainly to understand how its api should be used. I think most of the coverage probably comes from the other tests, but still it would be nice, at a quick glance on a test to see how it works.

woile commented on docs/tutorials/tag_format.md:
TODO?

noirbizarre replied:
Yes, it is now 👍🏼

noirbizarre commented:
I'll add some examples. I already pushed a few changes/fixes.
It started with a use case I have on lot of projects at work (tag convention that changed) and while working on this, I realized the tag management logic spread on multiple modules, that many things depends on it but on different phases, used in and that it was difficult to change, so I introduced the `TagRules` object. It should make easier to work on this kind of topic later.

Anyway, I have some time today, I'll advance the documentation and rework and hope to be able to provide a new iteration

noirbizarre commented on commitizen/changelog_formats/markdown.py:
```diff
@@ -12,28 +12,11 @@ class Markdown(BaseFormat):
     RE_TITLE = re.compile(r"^(?P<level>#+) (?P<title>.*)$")
 
-    def parse_version_from_title(self, line: str) -> str | None:
+    def parse_version_from_title(self, line: str) -> tuple[str, str] | None:
```
Done: added a `VersionTag` object. That's much clearer 👌🏼

noirbizarre commented:
There it is, PR updated with: 
- a tutorial
- more docstrings and examples
- all comments applied

I noticed a few possible edge-cases (not introduced in this PR) I might handle after this PR.  Example: it is possible to have `tag_format = ${major}` and then, tree processing is broken, incremental changelog is broken, version ranging is broken...
It is logical, but I think it is worth documenting it, and I think that's why there were 2 different logics (that I kept as it is in this PR) for version parsing:
- 1 assuming that it's possible to have only `${major}` or `${major}.${minor}` as tag format
- 1 assuming version sorting, changelog rendering always requires the full version

One solution may be to only allow `${version}` in file format and support some extra tag format (which would in addition allow bumping `v3.2.1`, `v3.2` ad `v3` at the same time, which have some use cases like GitHub actions versioning). But that's another story

noirbizarre commented:
@woile @Lee-W, if this is ok for you, I would love to merge this one too as I have addressed all comments.
(Given this is not a small Pr, keeping it open leads to conflicts and rebases)

Lee-W commented:
> @woile @Lee-W, if this is ok for you, I would love to merge this one too as I have addressed all comments. (Given this is not a small Pr, keeping it open leads to conflicts and rebases)

I'll try to take a look tomorrow or Friday, but feel free to merge it without me 🙂

Lee-W commented:
My comment is non-blocking. Feel free to merge it or make changes if you deem necessary. Awesome PR. Thanks a lot!

noirbizarre commented on commitizen/changelog_formats/base.py:
```diff
@@ -28,15 +26,12 @@ def __init__(self, config: BaseConfig):
         self.tag_format = self.config.settings["tag_format"]
-
-    @property
-    def version_parser(self) -> Pattern:
```
No, it has been designed not to be and should be totally backward-compatible.
The behavior will only change when you start using the new settings.

noirbizarre commented:
Thanks ! 🙏🏼 
PR merged, I'll be reactive if any breakage is introduced (but it shouldn't given how it has been tested)

carlossg commented:
this PR causes `cz bump` to break when one tag is invalid, see #1375
